File: scripts/render_diagrams.ts
#!/usr/bin/env bun
/**
 * Render all .dot and .mmd source files to .svg in docs/diagrams/.
 *
 * Priority:
 *   1. .dot files  → graphviz dot (always reliable)
 *   2. .mmd files  → mmdc with front-matter strip as fallback
 *
 * Usage:
 *   bun run scripts/render_diagrams.ts
 */

import { readFileSync, writeFileSync, readdirSync, existsSync } from "fs";
import { execFile } from "child_process";
import { join } from "path";
import { rmSync } from "fs";

// ─── Config ────────────────────────────────────────────────────────────────

const DIAGRAMS_DIR = join(process.cwd(), "docs", "diagrams");

// ─── Helpers ───────────────────────────────────────────────────────────────

/** Run a command, return true on exit code 0. */
async function run(
  cmd: string[],
  opts?: { capture?: boolean; cwd?: string },
): Promise<boolean> {
  return new Promise((resolve) => {
    const child = execFile(cmd[0], cmd.slice(1), {
      cwd: opts?.cwd ?? process.cwd(),
    }, (err, _stdout, stderr) => {
      if (err && opts?.capture) process.stderr.write(stderr);
      resolve(err === null);
    });
    if (!opts?.capture) child.stdout?.on("data", () => {});
    if (!opts?.capture) child.stderr?.on("data", () => {});
  });
}

function addWhiteBg(svgPath: string): void {
  const c = readFileSync(svgPath, "utf8");
  if (c.includes('fill="white"')) return;
  const patched = c.replace(
    /(<svg[^>]+>\n)/,
    '$1<rect width="100%" height="100%" fill="white"/>\n',
  );
  writeFileSync(svgPath, patched);
}

// ─── Renderers ─────────────────────────────────────────────────────────────

async function renderDot(src: string, out: string): Promise<boolean> {
  const ok = await run(["dot", "-Tsvg", src, "-o", out]);
  if (ok) addWhiteBg(out);
  return ok;
}

async function renderMmd(src: string, out: string): Promise<boolean> {
  // Fast path: mmdc handles files with no front matter
  let ok = await run(["mmdc", "-i", src, "-o", out, "-t", "neutral", "-b", "white"]);
  if (ok) return true;

  // Fallback: strip YAML front matter, then render
  const extracted = extractMermaidBlock(src);
  if (extracted === null) return false;

  const tmp = "/tmp/mmd-render.mmd";
  writeFileSync(tmp, extracted);
  ok = await run(["mmdc", "-i", tmp, "-o", out, "-t", "neutral", "-b", "white"]);
  try { rmSync(tmp); } catch {}
  return ok;
}

function extractMermaidBlock(filePath: string): string | null {
  const content = readFileSync(filePath, "utf8");
  const match = content.match(/\x60\x60\x60mermaid\n([\s\S]*?)\n\x60\x60\x60/);
  return match ? match[1] : null;
}

// ─── Main ───────────────────────────────────────────────────────────────────

async function main() {
  console.log("Rendering diagrams...\n");

  const files = readdirSync(DIAGRAMS_DIR).filter((f) =>
    f.endsWith(".dot") || f.endsWith(".mmd")
  );

  if (files.length === 0) {
    console.log("No .dot or .mmd files found in docs/diagrams/.");
    return;
  }

  let count = 0;
  for (const file of files.sort()) {
    const ext = file.endsWith(".dot") ? ".dot" : ".mmd";
    const src = join(DIAGRAMS_DIR, file);
    const out = src.replace(ext, ".svg");
    const extLabel = ext === ".dot" ? "DOT" : "MMD";

    process.stdout.write(`  ${file} (${extLabel}) → ${file.replace(ext, ".svg")} ... `);

    const ok = ext === ".dot"
      ? await renderDot(src, out)
      : await renderMmd(src, out);

    if (ok) {
      const size = existsSync(out) ? readFileSync(out).byteLength : 0;
      console.log(`✓ ${size.toLocaleString()}B`);
      count++;
    } else {
      console.log("✗ FAILED");
    }
  }

  console.log(`\nDone. Generated ${count} SVG file(s).`);
}

main();

File: scripts/render_diagrams.py
#!/usr/bin/env python3
"""Render all .mmd Mermaid files to .svg in docs/diagrams/."""
import subprocess, os, glob, sys

DIAGRAMS_DIR = "docs/diagrams"
HELPER = "scripts/extract_mermaid.py"

def run(cmd, capture=False):
    result = subprocess.run(cmd, capture_output=capture, text=True)
    if capture and result.returncode != 0:
        sys.stderr.write(result.stderr)
    return result.returncode == 0

print("Rendering Mermaid diagrams...")
count = 0
for f in sorted(glob.glob(f"{DIAGRAMS_DIR}/*.mmd")):
    out = f.replace(".mmd", ".svg")
    name = os.path.basename(f)
    print(f"  {name} -> {os.path.basename(out)}")

    # Try direct render first (fast path — works when no front matter)
    ok = run(["mmdc", "-i", f, "-o", out, "-t", "neutral", "-b", "white"])

    if not ok:
        # Fall back: strip YAML front matter with helper, then render
        tmp = "/tmp/mmd-render.mmd"
        if run(["python3", HELPER, f, tmp]):
            ok = run(["mmdc", "-i", tmp, "-o", out, "-t", "neutral", "-b", "white"])
            try:
                os.remove(tmp)
            except OSError:
                pass

    if ok:
        size = os.path.getsize(out)
        print(f"    ✓ {size:,} bytes")
        count += 1
    else:
        print(f"    ✗ FAILED")

print(f"\nDone. Generated {count} SVG file(s).")